lib/index.js

```diff
@@ -107,23 +115,47 @@ api.set = async ({
   } else if(type === 'nonce') {
     // FIXME: may need to allow up to N tokens, one per `clientId` if
     // `clientId` is set
-
-    // if there is an existing, unexpired nonce, reuse it, do not generate
-    // a new one to avoid DoS attack of constantly running bcrypt algorithm
+    const cfg = config['authn-token'];
+    const {maxNonceCount} = cfg.nonce.defaults;
+    let existingTokens;
     try {
-      const existingToken = await api.get({actor: null, account, email, type});
-      // keep existing token if it hasn't expired; note that there is no
-      // `challenge` to return because it is not known in the database
-      if(existingToken && new Date() < new Date(existingToken.expires)) {
-        return {type};
-      }
+      // check if nonce exists
+      existingTokens = await api.getAll({actor: null, account, email, type});
     } catch(e) {
       if(e.name !== 'NotFoundError') {
         throw e;
       }
     }
-
-    const cfg = config['authn-token'];
+    let removeCount = 0;
+    if(existingTokens) {
+      const now = new Date();
+      for(const existingToken of existingTokens) {
+        // check if existing token has expired
+        if(existingToken && now > new Date(existingToken.expires)) {
```

Comment:
Would this be an acceptable change? It removes the potentially unnecessary Date parsing, etc.

```suggestion
      const now = Date.now();
      for(const existingToken of existingTokens) {
        // check if existing token has expired
        if(existingToken && now > existingToken.expires) {
```

Reply:
Updated.